File: src/scanthosegates/ModPlugin.java
package scanthosegates;

import com.fs.starfarer.api.BaseModPlugin;
import com.fs.starfarer.api.Global;
import com.fs.starfarer.api.campaign.CharacterDataAPI;
import com.fs.starfarer.api.campaign.rules.MemoryAPI;
import lunalib.lunaSettings.LunaSettings;
import org.apache.log4j.Level;
import org.apache.log4j.Logger;
import scanthosegates.campaign.econ.abilities.CryosleeperScanner;
import scanthosegates.campaign.econ.abilities.GateScanner;
import scanthosegates.campaign.econ.abilities.HypershuntScanner;
import scanthosegates.campaign.listeners.RelocationListener;
import scanthosegates.campaign.listeners.SalvagingListener;

import java.util.MissingResourceException;

public class ModPlugin extends BaseModPlugin {
    private static final Logger log = Global.getLogger(ModPlugin.class);
    static {log.setLevel(Level.ALL);}
    public static final String ID = "scan_those_gates";
    public static final String PREFIX = "stg_";
    static final String LUNALIB_ID = "lunalib";
    public static final String INTEL_MEGASTRUCTURES = "Megastructures";
    static <T> T get(String id, Class<T> type) throws Exception {
        if (Global.getSettings().getModManager().isModEnabled(LUNALIB_ID)) {
            if (type == Boolean.class) return type.cast(LunaSettings.getBoolean(ModPlugin.ID, PREFIX + id));
        } else {
            if (type == Boolean.class) return type.cast(Global.getSettings().getBoolean(id));
        }
        throw new MissingResourceException("No setting found with id: " + id, type.getName(), id);
    }
    static boolean getBoolean(String id) throws Exception { return get(id, Boolean.class); }
    static void readSettings() {
        try {
            RevealAllGates = getBoolean("RevealInactiveGates");
            ActivateAllGates = getBoolean("ActivateInactiveGates");
        } catch (Exception e) {
            log.debug("Failed to read lunaSettings. Exception: " + e);
        }
    }
    public static boolean
            RevealAllGates = false,
            ActivateAllGates = false;

    @Override
    public void onGameLoad(boolean newGame){
        MemoryAPI sectorMemory = Global.getSector().getMemoryWithoutUpdate();
        CharacterDataAPI characterData = Global.getSector().getCharacterData();

        if (!sectorMemory.contains(GateScanner.UNSCANNED_GATES)) {
            sectorMemory.set(GateScanner.UNSCANNED_GATES, true);
        }
        if (!sectorMemory.contains(HypershuntScanner.CAN_SCAN_HYPERSHUNTS)) {
            sectorMemory.set(HypershuntScanner.CAN_SCAN_HYPERSHUNTS, true);
        }
        if (!sectorMemory.contains(CryosleeperScanner.CAN_SCAN_CRYOSLEEPERS)) {
            sectorMemory.set(CryosleeperScanner.CAN_SCAN_CRYOSLEEPERS, true);
        }

        if (!characterData.getAbilities().contains("stg_GateScanner")) {
            characterData.addAbility("stg_GateScanner");
        }
        if (!characterData.getAbilities().contains("stg_HypershuntScanner")) {
            characterData.addAbility("stg_HypershuntScanner");
        }
        if (!characterData.getAbilities().contains("stg_CryosleeperScanner")) {
            characterData.addAbility("stg_CryosleeperScanner");
        }

        Global.getSector().getListenerManager().addListener(new RelocationListener(), true);
        Global.getSector().getListenerManager().addListener(new SalvagingListener(), true);
    }

    @Override
    public void onApplicationLoad() {
        if (Global.getSettings().getModManager().isModEnabled(LUNALIB_ID)) {
            LunaSettings.addSettingsListener(new LunaSettingsChangedListener());
        }
    }
}

File: src/org/aero/scanThoseGates/ModPlugin.kt
package org.aero.scanThoseGates

import com.fs.starfarer.api.BaseModPlugin
import com.fs.starfarer.api.Global
import lunalib.lunaSettings.LunaSettings
import org.aero.scanThoseGates.campaign.abilities.GateScanner
import org.aero.scanThoseGates.campaign.listeners.RelocationListener
import org.aero.scanThoseGates.campaign.listeners.SalvagingListener
import org.aero.scanThoseGates.misc.LunaSettingsManager.addToManagerIfNeeded
import org.apache.log4j.Level

class ModPlugin : BaseModPlugin() {

    companion object Settings {
        private val log = Global.getLogger(ModPlugin::class.java)
        init { log.level = Level.ALL }

        const val ID = "scan_those_gates"
        const val LUNALIB_ID = "lunalib"

        val LUNALIB_ENABLED = Global.getSettings().modManager.isModEnabled(LUNALIB_ID)

        const val LUNA_MAJOR = 1
        const val LUNA_MINOR = 7
        const val LUNA_PATCH = 4

        var revealAllGates = false
        var activateAllGates = false

        var usableCheckInterval = 1f
        var usableCheckReportInterval = 60f

        const val GATE_SCAN_ABILITY = "stg_GateScanner"
        const val HYPERSHUNT_SCAN_ABILITY = "stg_HypershuntScanner"
        const val CRYOSLEEPER_SCAN_ABILITY = "stg_CryosleeperScanner"

        const val INTEL_MEGASTRUCTURES = "Megastructures"
        const val UNSCANNED_GATES = "\$UnscannedGatesFound"
        const val ALLOW_CRYOSLEEPER_SCAN = "\$CryosleeperScannerAllowed"
        const val ALLOW_HYPERSHUNT_SCAN = "\$HypershuntScannerAllowed"

        private fun getBoolean(id: String): Boolean {
            return if (LUNALIB_ENABLED) (LunaSettings.getBoolean(ID, id) ?: false) else Global.getSettings().getBoolean(id)
        }

        private fun getFloat(id: String, defaultValue: Float): Float {
            return if (LUNALIB_ENABLED) (LunaSettings.getFloat(ID, id) ?: defaultValue) else Global.getSettings().getFloat(id)
        }

        fun readSettings() {
            revealAllGates = getBoolean("RevealInactiveGates")
            activateAllGates = getBoolean("ActivateInactiveGates")

            usableCheckInterval = getFloat("GateScanUsableCheckInterval", 1f)
            usableCheckReportInterval = getFloat("GateScanDiagnosticsInterval", 60f)
        }
    }

    override fun onGameLoad(newGame: Boolean) {
        val sectorMemory = Global.getSector().memoryWithoutUpdate
        val characterData = Global.getSector().characterData
        val playerAbilities = characterData.abilities + Global.getSector().playerPerson.stats.grantedAbilityIds

        if (UNSCANNED_GATES !in sectorMemory) {
            sectorMemory[UNSCANNED_GATES] = false
        }
        if (ALLOW_HYPERSHUNT_SCAN !in sectorMemory) {
            sectorMemory[ALLOW_HYPERSHUNT_SCAN] = true
        }
        if (ALLOW_CRYOSLEEPER_SCAN !in sectorMemory) {
            sectorMemory[ALLOW_CRYOSLEEPER_SCAN] = true
        }

        if (GATE_SCAN_ABILITY !in playerAbilities) {
            characterData.addAbility(GATE_SCAN_ABILITY)
        }
        if (HYPERSHUNT_SCAN_ABILITY !in playerAbilities) {
            characterData.addAbility(HYPERSHUNT_SCAN_ABILITY)
        }
        if (CRYOSLEEPER_SCAN_ABILITY !in playerAbilities) {
            characterData.addAbility(CRYOSLEEPER_SCAN_ABILITY)
        }

        Global.getSector().listenerManager.addListener(RelocationListener(), true)
        Global.getSector().listenerManager.addListener(SalvagingListener(), true)

        readSettings()

        GateScanner.secondsSinceLastCheck = usableCheckInterval
        GateScanner.secondsSinceLastReport = usableCheckReportInterval
    }

    override fun onApplicationLoad() {
        if (LUNALIB_ENABLED) {
            if (requiredLunaLibVersionPresent()) {
                addToManagerIfNeeded()
            } else {
                throw RuntimeException("Using LunaLib with ScanThoseGates requires at least version " +
                        "$LUNA_MAJOR.$LUNA_MINOR.$LUNA_PATCH of LunaLib. Update your LunaLib."
                )
            }
        }
    }

    private fun requiredLunaLibVersionPresent(): Boolean {
        val version = Global.getSettings().modManager.getModSpec(LUNALIB_ID).version
        log.info("LunaLib Version: $version")
        val splitVersion = version.split("-".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()[0]
            .split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()

        val installedMajor = splitVersion[0].toInt()
        val installedMinor = splitVersion[1].toInt()
        val installedPatch = splitVersion[2].toInt()

        if (installedMajor > LUNA_MAJOR) {
            return true
        } else if (installedMajor < LUNA_MAJOR) {
            return false
        }
        if (installedMinor > LUNA_MINOR) {
            return true
        }
        return (installedMinor >= LUNA_MINOR && installedPatch >= LUNA_PATCH)
    }

}
